pkgs/development/tools/analysis/binlore/default.nix

```diff
@@ -56,58 +56,160 @@ let
+      fi
+      '' + lib.optionalString (builtins.hasAttr "binlore" drv) ''
+      if [[ -f "${drv.binlore}/$lore_type" ]]; then
+        cat "${drv.binlore}/$lore_type" >> "$out/$lore_type"
+      fi
+      '' + ''
+    done
 
     echo binlore for $drv written to $out
   '');
+
+  /*
+    Utility function for creating override lore for drv.
+
+    In the normal case, we attach this lore to `drv.passthru.binlore`.
+
+    > *Note*: We can reconsider the passthru attr name if someone adds
+    > a new lore provider. We settled on `.binlore` for now to make it
+    > easier for people to figure out what this is for.
+
+    The lore argument should be a Shell script (string) that generates
+    the necessary lore. You can use arbitrary Shell, but this function
+    includes a shell DSL you can use to declare/generate lore in most
+    cases. It has the following functions:
+
+    - `execer <verdict> [<path>...]`
+    - `wrapper <wrapper_path> <original_path>`
+
+    Writing every override explicitly in a Nix list would be tedious
+    for large packages, but this small shell DSL enables us to express
+    many overrides efficiently via pathname expansion/globbing.
+
+    Here's a very general example of both functions:
+
+    passthru.binlore = (binlore.synthesize finalAttrs.finalPackage ''
+      execer can bin/hello bin/{a,b,c}
+      wrapper bin/hello bin/.hello-wrapped
+    '');
+
+    And here's a specific example of how pathname expansion enables us
+    to express lore for the single-binary variant of coreutils while
+    being both explicit and (somewhat) efficient:
+
+    passthru = {} // optionalAttrs (singleBinary != false) {
+      binlore = (binlore.synthesize coreutils ''
+        execer can bin/{chroot,env,install,nice,nohup,runcon,sort,split,stdbuf,timeout}
+        execer cannot bin/{[,b2sum,base32,base64,basename,basenc,cat,chcon,chgrp,chmod,chown,cksum,comm,cp,csplit,cut,date,dd,df,dir,dircolors,dirname,du,echo,expand,expr,factor,false,fmt,fold,groups,head,hostid,id,join,kill,link,ln,logname,ls,md5sum,mkdir,mkfifo,mknod,mktemp,mv,nl,nproc,numfmt,od,paste,pathchk,pinky,pr,printenv,printf,ptx,pwd,readlink,realpath,rm,rmdir,seq,sha1sum,sha224sum,sha256sum,sha384sum,sha512sum,shred,shuf,sleep,stat,stty,sum,sync,tac,tail,tee,test,touch,tr,true,truncate,tsort,tty,uname,unexpand,uniq,unlink,uptime,users,vdir,wc,who,whoami,yes}
+      '');
+    };
+
+    Caution: Be thoughtful about using a bare wildcard (*) glob here.
+    We should generally override lore only when a human understands if
+    the executable will exec arbitrary user-passed executables. A bare
+    glob can match new executables added in future package versions
+    before anyone can audit them.
+  */
+  synthesize = drv: lore: runCommand "${drv.name}-lore-override" {
+    drv = drv;
+  } (''
+    execer(){
+      local verdict="$1"
+
+      shift
+
+      for path in "$@"; do
+        if [[ -e "$PWD/$path" ]]; then
```

Comment:
Are directories supported intentionally?

Reply:
Yes and no? :)

I couldn't remember when I was writing it whether the "regular file" bit of `-f` would exclude anything we would actually want to match here. I've had trouble in the past finding a lucid definition of what counts as a regular file and how symlink resolution fits in to the specific context of file property testing (POSIX's own definition strikes me as obtuse). Since this check is mostly just here to make sure we catch typos and removed paths, I decided overmatching was the smaller crime.

I'm happy to change it if you're confident `-f` won't leave out any edge cases we need?

Reply:
Marking resolved since I've ~optimistically changed this in the latest force-push. Can always revisit if it goes awry in some case.